u_contact helper

dy2403 opened the pull request:
Hi @sbillinge, can you please review my u_contact helper?

sbillinge commented on regolith/schemas.py:
```diff
@@ -1131,6 +1132,11 @@
             "description": "other names for the person",
         },
+        'day': {
+            "description": "day when the entry was created/updated",
```
which?  Created, or updated?  I think we want this to be the creation date, which is the first time we recognized that we knew this persion.  We have an updated field that can keep track of the most recent update.

dy2403 replied:
done!

sbillinge commented on regolith/schemas.py:
```diff
@@ -1164,6 +1175,21 @@
             "type": "string",
         },
+        'updated': {
+            "description": "date and time when the entry was created/updated",
```
this is the "most recently updated", not the created date

dy2403 replied:
done!

sbillinge commented on regolith/schemas.py:
```diff
@@ -1164,6 +1175,21 @@
+            "required": False,
+            "anyof_type": ["string", "datetime", "date"],
+        },
+        'year': {
```
please could you also add a 'date' field that duplicates the year/month/date info, but allows us to input an ISO date if we want.  I am moving to ISO dates in everything so this is a good chance to update things in the contacts collection

dy2403 replied:
hi @sbillinge, I modified it. Can you check if is that what you meant?

sbillinge replied:
yes, good

sbillinge commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,100 @@
+    def db_updater(self):
+        rc = self.rc
+        name = HumanName(rc.name)
+        day = dt.date.today().day
```
you don't need day, month and year here.  If you make ``now=dt.date.today()`` (I think that is the syntax) then ``now.year`` will get you ``year``

dy2403 replied:
done!

dy2403 commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,131 @@
+                                            "short form such as columbiau and will "
+                                            "be retrieved from institutions collection")
+    # subpi.add_argument("-e", "--email",
+    #                    help="email address")
```
Hi professor, @sbillinge
I could not add a field email, because it does not pass validators.py (ensure_email). 
Can you give me an idea of what I could do?

sbillinge replied:
an email field shouldn't be added to the db unless an email is specified at the command line, and then it mus be a valid email. ``something@thing.else`` so things should work fine.

dy2403 replied:
when I have a valid email (in this case, mine), it gives me the following error:
```
email = 'dy2403@barnard.edu-a'

    def ensure_email(email):
        """Ensures the email top-level key is well formed."""
>       email["url"] = ensure_string(email["url"])
E       TypeError: string indices must be integers
```

dy2403 replied:
added it as an issue in regolith (as we talked in the call).

sbillinge commented:
We would need to work on the schemas.  But there are some other schema
changes that are not merged in to regolith so I would like to go slowly on
this.  Maybe we could have a meeting with Michael (who is helping me with
these merges, and you and me and we can discuss?  Could you work with
Michael to arrange a meeting asap?

Thanks so much!

S

On Fri, Jun 5, 2020 at 1:01 PM dy2403 <notifications@github.com> wrote:

> *@dy2403* commented on this pull request.
> ------------------------------
>
> In regolith/helpers/u_contacthelper.py
> <https://github.com/regro/regolith/pull/390#discussion_r436048105>:
>
> > +import uuid
> +
> +from regolith.helpers.basehelper import DbHelperBase
> +from regolith.fsclient import _id_key
> +from regolith.tools import all_docs_from_collection
> +
> +
> +TARGET_COLL = "contacts"
> +
> +def subparser(subpi):
> +    subpi.add_argument("name", help="first name space last name in quotes")
> +    subpi.add_argument("institution", help="person's institution.  Can be inf"
> +                                            "short form such as columbiau and will "
> +                                            "be retrieved from institutions collection")
> +    # subpi.add_argument("-e", "--email",
> +    #                    help="email address")
>
> Hi professor, @sbillinge <https://github.com/sbillinge>
> I could not add a field email, because it does not pass validators.py
> (ensure_email).
> Can you give me an idea of what I could do?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/390#pullrequestreview-425471551>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOD3EVFNVZWAZF5633RVEQHNANCNFSM4NS7F75Q>
> .
>


-- 
Professor Simon Billinge
Columbia University

dy2403 commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,131 @@
+                       help="The date when the contact was created in ISO format"
+                            " Defaults to today's date."
+                       )
+    subpi.add_argument("--update",
```
ok, Professor! I added them there so I could pass the tests, otherwise it generates a new uuid and updated date and time. Do you have any solution for that?

sbillinge replied:
ah, good point.  Let me think about it.

sbillinge replied:
here is the answer for the uuid/date.  It is a bit tricky, it requires you to "mock" the uuid generation in the test:
https://stackoverflow.com/questions/44395755/how-to-mock-uuid-generation-in-a-test-case
https://stackoverflow.com/questions/23988853/how-to-mock-set-system-date-in-pytest

Have a quick play around with it, but if it is too difficult to understand, make an issue to fix that in your u_contact helper and we can merge it without that and someone else can take it on.

dy2403 commented:
Hi @sbillinge ! All I did was adding a news and changing the email to fixme. Still not sure of what I should do about --update and --uuid, as I need them there to pass the tests...

dy2403 commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,122 @@
+    #                    help="email address")
+    subpi.add_argument("-a", "--aliases", nargs='+',
+                        help="All the different ways that the person may "
+                             "be referred to as.  As many as you like, in "
```
you are right, professor! done!

dy2403 commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,122 @@
+                       help="The date and time when the contact was updated"
+                            " Defaults to the current date and time."
+                       )
+    subpi.add_argument("--uuid",
```
done!

dy2403 commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,131 @@
+            pdoc.update({'day': now.day, 'month': now.month, 'year': now.year, "date": now})
+            notes = []
+            aliases = [rc.name]
+            if not rc.uuid:
```
done!

dy2403 commented:
Hi @sbillinge !
I fixed the issues you pointed out! 
I changed the test to update the field, instead of adding a new contact. 
There is a failing test in test_builders.py, but this problem seem to be happening to other people in the group as well. Not sure how to solve it..

sbillinge commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,107 @@
+        else:
+            pdoc.update({'day': now.day, 'month': now.month, 'year': now.year, "date": dt.today()})
+            notes = []
+            aliases = [rc.name]
```
this can be empty list.  No need to repeat name in two places.

dy2403 replied:
done!

sbillinge commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,107 @@
+            notes = current.get('notes', [])
+            aliases = current.get('aka', [])
+        else:
+            pdoc.update({'day': now.day, 'month': now.month, 'year': now.year, "date": dt.today()})
```
just need date here.   no day/month/year which is for backward  compatibility.

dy2403 replied:
done!

sbillinge commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,107 @@
+            pdoc.update({'day': now.day, 'month': now.month, 'year': now.year, "date": dt.today()})
+            notes = []
+            aliases = [rc.name]
+            uid = str(uuid.uuid4())
```
please check, is key spelled right?  Though if we don't want this because we are not testing it, just leave the code out.

dy2403 replied:
hmm I changed the variable to UUID, is that what you meant?

sbillinge commented on regolith/helpers/u_contacthelper.py:
```diff
@@ -0,0 +1,107 @@
+        #FIXME
+        # if rc.e_email:
+        #     pdoc.update({"email": rc.email})
+        if rc.aliases:
```
please check, if aliases or notes are strings, this may not have the desired behavior (turn them into lists before doing this).

dy2403 replied:
I checked the schemas, and rc.notes can be either a list or a string, while rc.aliases can only be a list. 
So, here I added to check if rc.notes is a string.

sbillinge commented on regolith/schemas.py:
```diff
@@ -86,18 +86,17 @@
         "institution": "columbiau",
         "name": "Anthony B Friend",
-        "notes": "The guy I meet for coffee sometimes",
+        "notes": ["The guy I meet for coffee sometimes"],
```
check whether this can be a string or a list in the schema.   If it can be either we need to modify the code to handle either.

dy2403 replied:
up there!

dy2403 commented:
> great work. We may want a test that tests an add and one that tests an update, but the more tricky one is the update, so make sure your test tests the update.

ok, here I am updating the one that gets built from schemas, is that right? 
I am not sure how I could add a test for when we add a contact, as it seems that when I have two tests, I can't make both pass...

sbillinge commented:
> 
> 
> > great work. We may want a test that tests an add and one that tests an update, but the more tricky one is the update, so make sure your test tests the update.
> 
> ok, here I am updating the one that gets built from schemas, is that right?
> I am not sure how I could add a test for when we add a contact, as it seems that when I have two tests, I can't make both pass...

yes, this is correct.  There is an issue in regolith to fix this, but for now we can only do one test.  If you are testing already the updater, that is all I was asking for, so good.

sbillinge commented:
congrats @dy2403 !  Merged!
